esp-hal/src/spi/slave.rs

```diff
@@ -423,32 +407,45 @@ where
         self.enable_dma();
 
         reset_dma_before_load_dma_dscr(reg_block);
-        tx.prepare_transfer_without_start(
-            self.dma_peripheral(),
-            false,
-            write_buffer_ptr,
-            write_buffer_len,
-        )?;
+
         rx.prepare_transfer_without_start(
             false,
             self.dma_peripheral(),
             read_buffer_ptr,
             read_buffer_len,
         )?;
+        #[cfg(esp32s2)]
+        rx.start_transfer()?;
+
+        tx.prepare_transfer_without_start(
+            self.dma_peripheral(),
+            false,
+            write_buffer_ptr,
+            write_buffer_len,
+        )?;
+        #[cfg(esp32s2)]
+        tx.start_transfer()?;
 
-        self.clear_dma_interrupts();
         reset_dma_before_usr_cmd(reg_block);
 
-        // On the esp32, all full-duplex transfers are single, and all half-duplex
+        // Note: On the esp32, all full-duplex transfers are single, and all half-duplex
         // transfers use the cmd/addr/dummy/data sequence (but are still
         // single).
-        #[cfg(not(esp32))]
         reg_block
             .dma_conf()
             .modify(|_, w| w.dma_slv_seg_trans_en().clear_bit());
 
-        tx.start_transfer()?;
-        Ok(rx.start_transfer()?)
+        self.clear_dma_interrupts();
+        self.setup_for_flush();
+        reg_block.cmd().modify(|_, w| w.usr().set_bit());
+
+        #[cfg(not(esp32s2))]
```

Comment:
Could we get some explanation as to why all the chips apart from esp32s2 are started here, but esp32s2 is started earlier? This is quite hard to follow when looking at the driver for the first time.

Reply:
It was due to ESP32, it required special order of calls - it's not needed for S2, so I removed it